src/java/org/apache/cassandra/index/sai/disk/format/ComponentGroup.java

```diff
@@ -0,0 +1,264 @@
+ *     <li>the version the components are using. All the components of a group are on the same version</li>
+ *     <li>the generation, within the version, of the group/components. Generations are used when
+ *     {@link CassandraRelevantProperties#IMMUTABLE_SAI_COMPONENTS} is used, to avoid new builds to override (and thus
+ *     mutate) older builds. When the option of immutable components is not used, then the generation is fixed to 0.
+ *     See below for more details.</li>
+ * </ul>
+ * <p>
+ *
+ * <h2>Immutable components</h2>
+ *
+ * As mentioned above, when {@link CassandraRelevantProperties#IMMUTABLE_SAI_COMPONENTS} is enabled, existing components
+ * are not overwritten by new builds (the underlying intent is to allow rebuilding without stopping reads (to the
+ * rebuilt indexes). But unless a rebuild uses a different version than the existing components, the new components
+ * need "something" to distinguish them from the old ones, and it is what the generation provides.
+ * <p>
+ * The generation is specific to a component group, meaning that all the components of a group share the same
+ * generation (or to put it another way, if 2 components only differ by their generation, then they belong to different
+ * build and thus different groups), but that different groups can have different generations. For instance, if a
+ * specific index is rebuilt but without rebuilding the per-sstable components of each sstable, then after that rebuild
+ * the per-sstable groups will have generation 0, but the index groups will have generation 1.
+ * <p>
+ * When a sstable is "loaded", the "active" set of components to use is based on finding, for each kind of groups
+ * (per-sstable and per-index), the "complete" (see next paragraph) set of components with the highest version, and
+ * highest generation within that version.
+ * <p>
+ * A group of components may or may not be "complete" ({@link #isComplete()}): it is complete if the completion marker
+ * component for that group is present. Groups are temporarily incomplete during writing, but can also be more
+ * permanently incomplete for 2 main reasons: a build may fail mid-way, leaving one or more group incomplete, or we can
+ * have some corruption of the component files of some sort (corruption could here mean that a file is mistakenly
+ * deleted, or that the content of the file is corrupted somehow; the later triggers a removal of the corrupted file and
+ * of the completion marker, but not of the other component of the group). The bumping of generations takes incomplete
+ * groups into account, and so incomplete groups are not overridden either. Essentially, the generation used by a new
+ * build is always one more than the highest generation of any component found on disk (for the group in question, and
+ * the version we writting, usually {@link Version#LATEST}).
+ */
+public interface ComponentGroup
```

Comment:
Yeah, I guess the use of *group" is maybe a bit risky given `Index.Group`.

For variables, I think I'll just move to using `components`, or maybe `perSSTableComponents`/`perIndexComponents` when it feels like it matters.

For the class name, I guess `IndexComponents` make some sense, especially if I rename `IndexComponentInfo` to just `IndexComponent`. I do like that "group" carries the notion that it's not a completely random set of components, but not really a big deal.

I'll work out those renaming.

Reply:
Yeah, I guess the use of *group" is maybe a bit risky given `Index.Group`.

For variables, I think I'll just move to using `components`, or maybe `perSSTableComponents`/`perIndexComponents` when it feels like it matters.

For the class name, I guess `IndexComponents` make some sense, especially if I rename `IndexComponentInfo` to just `IndexComponent`. I do like that "group" carries the notion that it's not a completely random set of components, but not really a big deal.

I'll work out those renaming.

Reply:
Yeah, I guess the use of *group" is maybe a bit risky given `Index.Group`.

For variables, I think I'll just move to using `components`, or maybe `perSSTableComponents`/`perIndexComponents` when it feels like it matters.

For the class name, I guess `IndexComponents` make some sense, especially if I rename `IndexComponentInfo` to just `IndexComponent`. I do like that "group" carries the notion that it's not a completely random set of components, but not really a big deal.

I'll work out those renaming.

Reply:
I've pushed a bunch of renamings along the lines of what is above. There is no variable or type called "group", unless I missed something, in what is added, so hopefully that make it less confusing. I still reference to "group" in the documentation of `IndexComponents` and a couple internal places because that's still the cleanest way to explain it imo and I think it's clear from context anyway. Feel free to have a quick look and let me know if you have better ideas for some of the changes.